core/node/scrub/stream_scrub_task.go

```diff
@@ -0,0 +1,199 @@
+		scrubEventQueue: scrubEventQueue,
+		chainAuth:       chainAuth,
+		config:          cfg,
+	}
+	return proc, nil
+}
+
+func (tp *streamScrubTaskProcessorImpl) processTask(task *streamScrubTask) {
+	log := dlog.FromCtx(tp.ctx).With("Func", "streamScrubTask.process")
+	_, view, err := tp.cache.GetStream(tp.ctx, task.channelId)
+	if err != nil {
+		log.Error(
+			"Unable to scrub stream; could not fetch stream view",
+			"streamId",
+			task.channelId,
+			"error",
+			err,
+		)
+		return
+	}
+
+	members, err := view.(events.JoinableStreamView).GetChannelMembers()
+	if err != nil {
+		log.Error("Failed to fetch stream members", "error", err)
+		return
+	}
+	for member := range (*members).Iter() {
+		isEntitled, err := tp.chainAuth.IsEntitled(
+			tp.ctx,
+			tp.config,
+			auth.NewChainAuthArgsForChannel(
+				task.spaceId,
+				task.channelId,
+				member,
+				auth.PermissionRead,
+			),
+		)
+		if err != nil {
+			log.Error("Scrubbing error: unable to evaluate user entitlement",
+				"user",
+				member,
+				"channel",
+				task.channelId,
+				"spaceId",
+				task.spaceId,
+				"error",
+				err,
+			)
+			continue
+		}
+		// In the case that the user is not entitled, they must have lost their entitlement
+		// after joining the channel, so let's go ahead and boot them.
+		if !isEntitled {
+			userId, err := shared.AddressFromUserId(member)
+			if err != nil {
+				log.Error("Error converting user id into address", "member", member, "error", err)
+				continue
+			}
+			userStreamId, err := shared.UserStreamIdFromBytes(userId)
+			if err != nil {
+				log.Error(
+					"Error constructing user id stream from user address",
+					"userAddress",
+					userId,
+					"error",
+					err,
+				)
+			}
+
+			log.Info("Entitlement loss detected; enqueueing scrub for user",
+				"user",
+				member,
+				"userStreamId",
+				userStreamId,
+				"channel",
+				task.channelId,
+				"space",
+				task.spaceId,
+			)
+			tp.scrubEventQueue <- &rules.DerivedEvent{
+				StreamId: userStreamId,
+				Payload: events.Make_UserPayload_Membership(
+					protocol.MembershipOp_SO_LEAVE,
+					task.channelId,
+					&member,
+					task.spaceId[:],
+				),
+			}
+		}
+	}
+}
+
+type streamScrubTask struct {
+	channelId     shared.StreamId
+	spaceId       shared.StreamId
+	taskProcessor *streamScrubTaskProcessorImpl
+}
+
+func (t *streamScrubTask) process() {
+	t.taskProcessor.processTask(t)
+}
+
+// TryScheduleScrub schedules a stream scrub task if:
+// - the stream is a channel stream,
+// - the stream has not been recently scrubbed, and
+// - there is no pending scrub for the given stream.
+// If the worker pool is full, the method will not block but will return an error.
+// This is so that we don't affect ability to post updates to channels with stale scrubs
+// whenever the node falls behind due to high scrubbing volume.
+func (tp *streamScrubTaskProcessorImpl) TryScheduleScrub(
+	ctx context.Context,
+	streamId shared.StreamId,
+) (bool, error) {
```

Comment:
The return values are ignored everywhere should we not just drop them and do any logging/future metrics in this function?

Reply:
I added an unused return value to make this forward compatible with future uses of scrubbing that we may want to error on, say if we expose an RPC endpoint in our API. We're not implementing that at this time, but we've discussed it as an option, so I kept that in mind as a future possibility when designing.

At this time, my strategy was to *also* aggressively log errors from `TryScheduleScrub` since we don't want to raise an error from any current call sites if this fails.

Reply:
Metrics yes! I really want those but wanted to do that in a follow up PR.